Kubelet: support retrieving stats using UID of mirror pods

yujuhong opened the pull request:
Kubelet supports retrieving stats for pods/containers with and without UID.
This does not always work for the static pods because users may get the UIDs of
the mirror pods from the API server, and use them to query Kubelet. In this
case, Kubelet would fail to locate the containers due to mismatched UIDs.

This change adds a intenral mirror to static pod UID mapping and teaches all
public-facing functions to perform UID lookup before proceeding. This allows
users to use either mirror or static pod's UID to retrieve stats.

This fixes #5688

yujuhong committed:
Kubelet: support retrieving stats using UID of mirror pod

Kubelet supports retrieving stats for pods/containers with and without UID.
This does not always work for the static pods because users may get the UIDs of
the mirror pods from the API server, and use them to query Kubelet. In this
case, Kubelet would fail to locate the containers due to mismatched UIDs.

This change adds a intenral mirror to static pod UID mapping and teaches all
public-facing functions to perform UID lookup before proceeding. This allows
users to use either mirror or static pod's UID to retrieve stats.

yujuhong commented:
This PR is a bit cumbersome. The pod manager that I'm supposedly working on will help a lot.

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -2176,8 +2200,23 @@ func (kl *Kubelet) StreamingConnectionIdleTimeout() time.Duration {
 	return kl.streamingConnectionIdleTimeout
 }
 
+func (kl *Kubelet) getStaticPodUID(uid types.UID) (types.UID, bool) {
```
Lets document why this exists in case future readers get confused :)

yujuhong replied:
Done.

vmarmol commented on pkg/kubelet/kubelet.go:
```diff
@@ -2176,8 +2200,23 @@ func (kl *Kubelet) StreamingConnectionIdleTimeout() time.Duration {
+
+	// If the UID belongs to a mirror pod, translate it to the UID of its
+	// static counterpart.
+	if uid != "" {
```
WDYT of doing this logic in getStaticPodUID() since we do the same thing in all callsites? Maybe getPodUID()

yujuhong replied:
Done.

vmarmol commented on pkg/kubelet/mirror_manager.go:
```diff
@@ -123,16 +118,90 @@ func isMirrorPod(pod *api.Pod) bool {
+	return filteredPods, *mirrorPods
+}
+
+// This is not thread-safe to use alone.
```
nit: This is "thread-compatible". Internally we commonly use: thread-safe, thread-compatible, and thread-hostile. This article explains the definition of each: http://www.ibm.com/developerworks/library/j-jtp09263/

yujuhong replied:
Done. Thanks for the pointer!

vmarmol commented:
Yeah it does look a bit brittle, but I think we knew that coming into it. Looks good overall, minor comments. Thanks @yujuhong!

vmarmol commented:
LGTM, squash commits and we'll merge.

erictune commented:
What is this for?

On Fri, Mar 20, 2015 at 3:08 PM, Victor Marmol <notifications@github.com>
wrote:

> LGTM, squash commits and we'll merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5730#issuecomment-84166671>
> .
>

vmarmol commented:
@erictune fixing #5688. Mirror pods fail getting stats today because they have a different UID than their static equivalent. This does the conversion in Kubelet API requests.

erictune commented:
Who would get these stats?  If you say "master", then: isn't the new
direction to have kubelet post pod status to its master rather than having
the master poll for status?

On Fri, Mar 20, 2015 at 3:49 PM, Victor Marmol <notifications@github.com>
wrote:

> @erictune <https://github.com/erictune> fixing #5688
> <https://github.com/GoogleCloudPlatform/kubernetes/issues/5688>. Mirror
> pods fail getting stats today because they have a different UID than their
> static equivalent. This does the conversion in Kubelet API requests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/5730#issuecomment-84174116>
> .
>

vmarmol commented:
Heapster uses this stats API

yujuhong commented:
Squashed and rebased.

Shippable's complaints seem irrelevant though.
```
pkg/kubelet/container/fake_runtime.go:26:2: cannot find package "github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume" in any of:
	/home/shippable/.gvm/gos/go1.4/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume (from $GOROOT)
	/home/shippable/workspace/src/github.com/GoogleCloudPlatform/kubernetes/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume (from $GOPATH)
	/home/shippable/workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume
```

vmarmol commented:
@yujuhong yeah that seemed odd, I kicked it again and we'll see what it thinks.

yujuhong commented:
@erictune, heapster queries kubelet directly for container info.

vishh commented:
Thanks for the quick fix @yujuhong!

yujuhong commented:
The head is broken :\ Other PRs are suffering too. This seems to be relevant to #5642.

vmarmol commented:
:( @vishh do you know if anyone is working on a fix?

yujuhong commented:
I uploaded one: #5743

vmarmol commented:
Build is green! Merging.
